airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py

```diff
@@ -91,10 +92,12 @@ def read(
         logger: logging.Logger,
         config: Mapping[str, Any],
         catalog: ConfiguredAirbyteCatalog,
-        state: MutableMapping[str, Any] = None,
+        state: Union[List[AirbyteStateMessage], MutableMapping[str, Any]] = None,
     ) -> Iterator[AirbyteMessage]:
         """Implements the Read operation from the Airbyte Specification. See https://docs.airbyte.io/architecture/airbyte-protocol."""
-        connector_state = copy.deepcopy(state or {})
+        state_manager = ConnectorStateManager(state=state)
+        connector_state = copy.deepcopy(state_manager.get_legacy_state(), {})
```

Comment:
I wonder if we could make `ConnectorStateManagerget_legacy_state` return the deep copy.

Reply:
yeah that is a good point. thinking about it in the scope of the upcoming tickets that update/return the legacy state, it might be safer if we were to always return a deep copy so that we can only ever manipulate the underlying dictionary via the update_state_for_stream() method. will fix this!